File: src/blender_bridge.py
"""Drive headless Blender to convert a MeshChunk → .fbx.

We don't import bpy in this process. Instead we marshal the chunk's geometry
to JSON, spawn ``blender --background --python scripts/blender_export.py --
<chunk.json>``, and let Blender do the FBX export with its stock exporter.
That keeps the main app independent of Blender's Python and lets us upgrade
Blender freely.
"""
from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

import numpy as np

from chunker import MeshChunk
from fm4.ir import TrackIR
from materials import TM2020Material


def find_blender(override: Path | None = None) -> Path:
    """Locate the Blender binary. Tries override → Steam-Linux → PATH.

    Steam Blender on Linux installs at .../SteamLibrary/steamapps/common/Blender/blender;
    other paths can be supplied via ``override`` (Settings UI feeds this).
    """
    if override is not None:
        p = Path(override)
        if not p.is_file():
            raise FileNotFoundError(f"blender override path does not exist: {p}")
        return p

    candidates = [
        Path("/run/media/paths/SSS-Games/SteamLibrary/steamapps/common/Blender/blender"),
        Path.home() / ".steam/steam/steamapps/common/Blender/blender",
        Path("/usr/bin/blender"),
        Path("/opt/blender/blender"),
    ]
    for c in candidates:
        if c.is_file():
            return c

    import shutil
    found = shutil.which("blender")
    if found:
        return Path(found)

    raise FileNotFoundError(
        "Blender not found. Set the override path in Settings or install Steam Blender."
    )


def dump_chunk(
    chunk: MeshChunk,
    track: TrackIR,
    materials_by_chunk_mat: dict[tuple[int, int], TM2020Material],
    out_fbx: Path,
    out_json: Path,
) -> Path:
    """Serialize the chunk's geometry + material names + transforms to JSON.

    `materials_by_chunk_mat` maps (mesh_key, fm4_mat_index) → TM2020Material;
    we only need the .name for FBX, but the caller has already built them so
    we accept the full dict to avoid recomputation.
    """
    meshes_payload: dict[str, dict] = {}
    for mk in chunk.mesh_keys:
        m = track.meshes.get(mk)
        if m is None:
            continue
        verts = m.vertices.tolist()
        faces = m.faces.tolist()
        uvs = m.uvs.tolist() if m.uvs is not None else None
        mat_per_face = m.material_per_face.tolist()

        names: list[str] = []
        for mat_idx in range(len(m.materials)):
            tm = materials_by_chunk_mat.get((mk, mat_idx))
            names.append(tm.name if tm else f"mat_{mk:08x}_{mat_idx:03d}")

        meshes_payload[str(mk)] = {
            "verts": verts,
            "faces": faces,
            "uvs": uvs,
            "material_per_face": mat_per_face,
            "material_names": names,
        }

    instances_payload: list[dict] = []
    for inst in chunk.instances:
        base = inst.model_index << 8
        keys = [base | s for s in range(256) if (base | s) in track.meshes and (base | s) in chunk.mesh_keys]
        if not keys:
            continue
        instances_payload.append({
            "transform": np.asarray(inst.transform, dtype=np.float32).tolist(),
            "mesh_keys": [str(k) for k in keys],
        })

    payload = {
        "out_fbx": str(out_fbx),
        "meshes": meshes_payload,
        "instances": instances_payload,
    }
    out_json.parent.mkdir(parents=True, exist_ok=True)
    out_json.write_text(json.dumps(payload))
    return out_json


def export_chunk_to_fbx(
    chunk_json: Path,
    blender_path: Path,
    blender_export_script: Path,
    timeout: float = 600.0,
) -> None:
    """Spawn headless Blender to consume chunk_json and write the FBX.

    Raises RuntimeError on non-zero exit; surfaces stderr verbatim.
    """
    cmd = [
        str(blender_path),
        "--background",
        "--factory-startup",
        "--python", str(blender_export_script),
        "--",
        str(chunk_json),
    ]
    # stdin=DEVNULL: under PyInstaller --windowed on Windows, the parent
    # has no console so its stdin handle is invalid; subprocess inheriting
    # it fails with WinError 6. Explicitly nulling stdin avoids that.
    result = subprocess.run(
        cmd,
        stdin=subprocess.DEVNULL,
        capture_output=True,
        text=True,
        timeout=timeout,
    )
    if result.returncode != 0:
        raise RuntimeError(
            f"blender export failed (rc={result.returncode}):\n"
            f"--- stdout ---\n{result.stdout}\n"
            f"--- stderr ---\n{result.stderr}"
        )
    if "OK:" not in result.stdout:
        # Blender may exit 0 even on partial failures; sentinel-check the script
        print("warning: blender exited 0 but no OK sentinel found", file=sys.stderr)
        print(result.stdout, file=sys.stderr)

File: app/blender_bridge.nim
## Drive headless Blender to convert a MeshChunk → .fbx.
##
## Nim port of `src/blender_bridge.py`. We never import bpy: marshal the chunk's
## geometry to JSON, spawn `blender --background --factory-startup --python
## scripts/blender_export.py -- <chunk.json>`, and let Blender's stock FBX
## exporter do the work (blender_export.py runs INSIDE Blender, untouched). That
## keeps the orchestrator independent of Blender's Python.
##
## The Wine `.exe`-vs-ELF guard from the Python version is dropped — the native
## build never runs under Wine. Blender location reuses settings.detectBlender
## (the cross-platform Steam-library detector), with a Settings override on top.

import std/[os, json, tables, strutils, sets]
import ir, chunker, materials, subproc, settings

proc findBlender*(override = ""): string =
  ## Locate the Blender binary: override (Settings) → settings.detectBlender
  ## (Steam libs / Program Files / Unix dirs / PATH). Raises if none found.
  if override.len > 0:
    if not fileExists(override):
      raise newException(IOError, "blender override path does not exist: " & override)
    return override
  let found = detectBlender()
  if found.len == 0:
    raise newException(IOError,
      "Blender not found. Set the override path in Settings or install Blender.")
  return found

proc dumpChunk*(chunk: MeshChunk; track: TrackIR;
                materials: Table[(int, int), TM2020Material];
                outFbx, outJson: string): string =
  ## Serialize the chunk's geometry + material names + instance transforms to
  ## the JSON that blender_export.py consumes; return the JSON path. Mirrors
  ## dump_chunk in src/blender_bridge.py byte-semantically (floats round-trip).
  var chunkKeySet = initHashSet[int]()
  for k in chunk.meshKeys: chunkKeySet.incl(k)

  var meshesPayload = newJObject()
  for mk in chunk.meshKeys:
    if not track.meshes.hasKey(mk): continue
    let m = track.meshes[mk]

    var verts = newJArray()
    for v in m.vertices:
      verts.add(%*[v[0].float, v[1].float, v[2].float])
    var faces = newJArray()
    for fc in m.faces:
      faces.add(%*[fc[0].int, fc[1].int, fc[2].int])
    var uvs: JsonNode
    if m.hasUvs:
      uvs = newJArray()
      for u in m.uvs: uvs.add(%*[u[0].float, u[1].float])
    else:
      uvs = newJNull()
    var mpf = newJArray()
    for x in m.materialPerFace: mpf.add(%(x.int))

    var names = newJArray()
    for matIdx in 0 ..< m.materials.len:
      if materials.hasKey((mk, matIdx)):
        names.add(%materials[(mk, matIdx)].name)
      else:
        names.add(% ("mat_" & toHex(mk, 8).toLowerAscii & "_" & intToStr(matIdx, 3)))

    meshesPayload[$mk] = %*{
      "verts": verts,
      "faces": faces,
      "uvs": uvs,
      "material_per_face": mpf,
      "material_names": names,
    }

  var instancesPayload = newJArray()
  for inst in chunk.instances:
    let base = inst.modelIndex shl 8
    var keys: seq[int]
    for s in 0 ..< 256:
      let k = base or s
      if track.meshes.hasKey(k) and k in chunkKeySet: keys.add(k)
    if keys.len == 0: continue
    # Emit as a NESTED 4x4 (row-major), matching numpy's (4,4).tolist() — this
    # is what blender_export.py's `Matrix(inst["transform"])` expects. Mat4 is
    # the row-major flat 16, so row r is elements [r*4 .. r*4+3].
    var tf = newJArray()
    for r in 0 ..< 4:
      var row = newJArray()
      for c in 0 ..< 4: row.add(%(inst.transform[r*4 + c].float))
      tf.add(row)
    var mkeys = newJArray()
    for k in keys: mkeys.add(% $k)
    instancesPayload.add(%*{"transform": tf, "mesh_keys": mkeys})

  let payload = %*{
    "out_fbx": outFbx,
    "meshes": meshesPayload,
    "instances": instancesPayload,
  }
  createDir(parentDir(outJson))
  writeFile(outJson, $payload)
  return outJson

proc exportChunkToFbx*(chunkJson, blenderPath, blenderExportScript: string) =
  ## Spawn headless Blender to consume chunkJson and write the FBX. Raises on a
  ## non-zero exit; warns (no raise) if Blender exits 0 without the OK sentinel.
  let args = @["--background", "--factory-startup",
               "--python", blenderExportScript, "--", chunkJson]
  let res = runCaptured(blenderPath, args)
  if res.rc != 0:
    raise newException(IOError,
      "blender export failed (rc=" & $res.rc & "):\n" & res.output)
  if "OK:" notin res.output:
    stderr.writeLine("warning: blender exited 0 but no OK sentinel found")
    stderr.writeLine(res.output)
